### PR TITLE
US-8 Created login mechanics for Firebase and a mock userbase for testing purposes

### PR DESCRIPTION
**Pull request by HVYVince**
object user is the singleton holding the current logged user's info
IUserConnector is the interface a facade should have to be usable in authentication mechanics
MockUserConnector and its test suite allows for testing outside of firebase
FirebaseConnector is to be used in production

### CODE REVIEW

File: app/src/androidTest/java/ch/epfl/sdp/LoginActivityTest.kt [hunk 1]
```diff
@@ -0,0 +1,79 @@
...
+    val activityRule = ActivityTestRule(LoginActivity::class.java)
+
+    @Before
+    fun forceDisconnect() {
+        user.connected = false
+    }
+
+    @Test
+    fun dummyTest() {
+        Assert.assertEquals(4, 2 + 2.toLong())
```
Comment (ProtectedVariable) [inline app/src/androidTest/java/ch/epfl/sdp/LoginActivityTest.kt]: Not sure that's important

File: app/src/test/java/ch/epfl/sdp/ExampleUnitTest.kt [hunk 2]
```diff
@@ -0,0 +1,16 @@
...
+
+/**
+ * Example local unit test, which will execute on the development machine (host).
+ *
+ * @see [Testing documentation](http://d.android.com/tools/testing)
+ */
+class ExampleUnitTest {
+    @Test
+    fun addition_isCorrect() {
+        Assert.assertEquals(4, 2 + 2.toLong())
```
Comment (ProtectedVariable) [inline app/src/test/java/ch/epfl/sdp/ExampleUnitTest.kt]: Not sure that's useful

**Review by ProtectedVariable — CHANGES_REQUESTED** [review]

File: app/src/test/java/ch/epfl/sdp/ExampleUnitTest.kt (hunk 2 above)
Comment (HVYVince) [reply app/src/test/java/ch/epfl/sdp/ExampleUnitTest.kt]: Yup, that will go

File: app/src/androidTest/java/ch/epfl/sdp/LoginActivityTest.kt (hunk 1 above)
Comment (HVYVince) [reply app/src/androidTest/java/ch/epfl/sdp/LoginActivityTest.kt]: Same, will change

**Review by ProtectedVariable — APPROVED** [review]

File: app/src/main/java/ch/epfl/sdp/MainActivity.kt [hunk 3]
```diff
@@ -0,0 +1,26 @@
...
+
+import android.content.Intent
+import android.os.Bundle
+import androidx.appcompat.app.AppCompatActivity
+import ch.epfl.sdp.databinding.ActivityMainBinding
+import ch.epfl.sdp.lobby.LobbyActivity
+
+class MainActivity : AppCompatActivity() {
+    private var _binding: ActivityMainBinding? = null
+    private val binding get() = _binding!!
```
Comment (Foebus) [inline app/src/main/java/ch/epfl/sdp/MainActivity.kt]: Prefer using a lateinit in the context of activities, it prevent from having nullable variables

File: app/src/main/java/ch/epfl/sdp/LoginActivity.kt [hunk 4]
```diff
@@ -0,0 +1,53 @@
...
+import android.text.Editable
+import androidx.appcompat.app.AppCompatActivity
+import ch.epfl.sdp.databinding.ActivityLoginBinding
+import kotlinx.android.synthetic.main.activity_login.*
+import ch.epfl.sdp.authentication.*
+
+class LoginActivity : AppCompatActivity() {
+
+    private var _binding: ActivityLoginBinding? = null
+    private val binding get() =_binding!!
```
Comment (Foebus) [inline app/src/main/java/ch/epfl/sdp/LoginActivity.kt]: Prefer using lateinit for activities, it avoid nullable variables

File: app/src/main/java/ch/epfl/sdp/LoginActivity.kt [hunk 5]
```diff
@@ -0,0 +1,53 @@
...
+
+    private fun signIn(email:Editable, password:Editable, connector: IUserConnector) {
+        connector.connect(email.toString(), password.toString())
+        if(user.connected)
+            changeDummyText("User logged in as " + user.username)
+        else
+            changeDummyText("Logging failed")
+    }
+
+    private fun changeDummyText(text:String) {
```
Comment (Foebus) [inline app/src/main/java/ch/epfl/sdp/LoginActivity.kt]: Are you sure about the name of that function?

File: app/src/androidTest/java/ch/epfl/sdp/authentication/MockUserTest.kt [hunk 6]
```diff
@@ -0,0 +1,71 @@
...
+    fun connectionWithWrongPasswordFails() {
+        val mock = MockUserConnector()
+        Assert.assertFalse(mock.connect("test0@test.com", "its2AMrightnow"))
+        Assert.assertFalse(user.connected)
+    }
+
+    @Test
+    fun canAddUserThenConnectToIt() {
+        val mock = MockUserConnector()
+        Assert.assertTrue(mock.register("testing", "C-C-C-CANDEAAAA"))
```
Comment (Foebus) [inline app/src/androidTest/java/ch/epfl/sdp/authentication/MockUserTest.kt:55]: Are you sure this test should be used in master?

**Review by Foebus — CHANGES_REQUESTED** [review]

File: app/src/main/java/ch/epfl/sdp/LoginActivity.kt (hunk 5 above)
Comment (HVYVince) [reply app/src/main/java/ch/epfl/sdp/LoginActivity.kt]: Yep, currently a dummy component in the Login activity has a text changing to acknowledge success of loggin

**Review by Foebus — APPROVED** [review]

**Pull request merged**